src/ray/core_worker/actor_manager.h

```diff
@@ -14,34 +14,122 @@
+class ActorManager {
  public:
-  ActorManager(gcs::ActorInfoAccessor &actor_accessor)
-      : actor_accessor_(actor_accessor) {}
+  explicit ActorManager(
+      std::shared_ptr<gcs::GcsClient> gcs_client,
+      std::shared_ptr<CoreWorkerDirectActorTaskSubmitterInterface> direct_actor_submitter,
+      std::shared_ptr<ReferenceCounterInterface> reference_counter)
+      : gcs_client_(gcs_client),
+        direct_actor_submitter_(direct_actor_submitter),
+        reference_counter_(reference_counter) {}
+
+  ~ActorManager() {}
+
+  friend class ActorManagerTest;
+
+  /// Register an actor handle.
+  ///
+  /// This should be called when an actor handle is given to us by another task
+  /// or actor. This may be called even if we already have a handle to the same
+  /// actor.
+  ///
+  /// \param[in] actor_handle The actor handle.
+  /// \param[in] outer_object_id The object ID that contained the serialized
+  /// actor handle, if any.
+  /// \param[in] caller_id The caller's task ID
+  /// \param[in] call_site The caller's site.
+  /// \return The ActorID of the deserialized handle.
+  ActorID RegisterActorHandle(std::unique_ptr<ActorHandle> actor_handle,
+                              const ObjectID &outer_object_id, const TaskID &caller_id,
+                              const std::string &call_site,
+                              const rpc::Address &caller_address);
+
+  /// Get a handle to an actor.
+  ///
+  /// \param[in] actor_id The actor handle to get.
+  /// \return reference to the actor_handle's pointer.
+  /// NOTE: Returned actorHandle should not be stored anywhere.
+  std::unique_ptr<ActorHandle> &GetActorHandle(const ActorID &actor_id);
+
+  /// Check if an actor handle that corresponds to an actor_id exists.
+  /// \param[in] actor_id The actor id of a handle.
+  /// \return True if the actor_handle for an actor_id exists. False otherwise.
+  bool CheckActorHandleExists(const ActorID &actor_id);
 
-  /// Called when an actor that we own can no longer be restarted.
-  void PublishTerminatedActor(const TaskSpecification &actor_creation_task) override;
+  /// Give this worker a handle to an actor.
+  ///
+  /// This handle will remain as long as the current actor or task is
+  /// executing, even if the Python handle goes out of scope. Tasks submitted
+  /// through this handle are guaranteed to execute in the same order in which
+  /// they are submitted.
+  ///
+  /// \param actor_handle The handle to the actor.
+  /// \param is_owner_handle Whether this is the owner's handle to the actor.
+  /// The owner is the creator of the actor and is responsible for telling the
+  /// actor to disconnect once all handles are out of scope.
+  /// \param[in] caller_id The caller's task ID
+  /// \param[in] call_site The caller's site.
+  /// \return True if the handle was added and False if we already had a handle
+  /// to the same actor.
+  bool AddActorHandle(std::unique_ptr<ActorHandle> actor_handle, bool is_owner_handle,
```

Comment:
If we make this a private method, we can remove the `bool is_owner_handle` from the public method's arguments (it should always be true when creating the actor).